Added PFDiJetCorrCheckerWithDiTau module to remove overlap between di-tau and di-jet for HLT

namppl opened the pull request:
We would like to introduce a new EDProducer "PFJetsTauCorrelationCondition" which will be used in VBF+2tau triggers [1]. The function of this module is quite similar to the recent PR for 2tau+1jet triggers [2]. The module looks for a combination of (j1, j2, t1, t2) (j1, j2 from PFJet, t1,t2 from the tau collection) that there is no matching between j1,j2 and t1,t2, i.e. they are 4 independent objects, and store j1 and j2.

[1] https://its.cern.ch/jira/browse/CMSHLT-2224
[2] https://github.com/cms-sw/cmssw/pull/36828

mbluj commented on RecoTauTag/HLTProducers/interface/PFJetsTauCorrelationCondition.h:
```diff
@@ -0,0 +1,34 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/L1Trigger/interface/Tau.h"
```
Not needed

mbluj commented on RecoTauTag/HLTProducers/interface/PFJetsTauCorrelationCondition.h:
```diff
@@ -0,0 +1,34 @@
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/L1Trigger/interface/Tau.h"
+#include "DataFormats/JetReco/interface/CaloJetCollection.h"
```
Not needed

mbluj commented on RecoTauTag/HLTProducers/interface/PFJetsTauCorrelationCondition.h:
```diff
@@ -0,0 +1,34 @@
+#include "DataFormats/TauReco/interface/PFTauFwd.h"
+#include "DataFormats/HLTReco/interface/TriggerFilterObjectWithRefs.h"
+#include "DataFormats/HLTReco/interface/TriggerObject.h"
+#include "DataFormats/HLTReco/interface/TriggerEvent.h"
```
Not needed

mbluj commented on RecoTauTag/HLTProducers/interface/PFJetsTauCorrelationCondition.h:
```diff
@@ -0,0 +1,34 @@
+#include "DataFormats/HLTReco/interface/TriggerObject.h"
+#include "DataFormats/HLTReco/interface/TriggerEvent.h"
+
+class PFJetsTauCorrelationCondition : public edm::global::EDProducer<> {
```
global->stream EDProducer, or there is any reason to use global?

mbluj commented on RecoTauTag/HLTProducers/interface/PFJetsTauCorrelationCondition.h:
```diff
@@ -0,0 +1,34 @@
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/L1Trigger/interface/Tau.h"
+#include "DataFormats/JetReco/interface/CaloJetCollection.h"
+#include "DataFormats/TauReco/interface/PFTauFwd.h"
```
Not needed

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+      pfJetSrc_(consumes<reco::PFJetCollection>(iConfig.getParameter<edm::InputTag>("PFJetSrc"))),
+      extraTauPtCut_(iConfig.getParameter<double>("extraTauPtCut")),
+      mjjMin_(iConfig.getParameter<double>("mjjMin")),
+      matchingR2_(iConfig.getParameter<double>("Min_dR") * iConfig.getParameter<double>("Min_dR")) {
```
Min_dR -> dRmin for consistency with other parameters
use `std:pow(iConfig.getParameter<double>("Min_dR"),2)`

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+
+                double mjj = (myPFJet1.p4() + myPFJet2.p4()).M();
+                if (mjj < mjjMin_) continue;
+
```
It is not mandatory to define mjj which is not used anywhere except this condition:
`if ((myPFJet1.p4() + myPFJet2.p4()).M() < mjjMin_) continue;`

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+                for (unsigned int iTau1 = 0; iTau1 < taus.size(); iTau1++) {
+                    bool isMatched11 = (reco::deltaR2(taus[iTau1]->p4(), myPFJet1.p4()) < matchingR2_);
+                    bool isMatched12 = (reco::deltaR2(taus[iTau1]->p4(), myPFJet2.p4()) < matchingR2_);
+                    if (isMatched11 || isMatched12) continue;
```
Check one condition after other to not evaluate dR w/o necessity:
```
if (reco::deltaR2(taus[iTau1]->p4(), myPFJet1.p4()) < matchingR2_) continue;
if (reco::deltaR2(taus[iTau1]->p4(), myPFJet2.p4()) < matchingR2_) continue;
```

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+                        bool isMatched21 = (reco::deltaR2(taus[iTau2]->p4(), myPFJet1.p4()) < matchingR2_);
+                        bool isMatched22 = (reco::deltaR2(taus[iTau2]->p4(), myPFJet2.p4()) < matchingR2_);
+                        if (isMatched21 || isMatched22) continue;
+
```
As above.

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+                    }
+                }
+
+                if (correctComb == true) {
```
`if (correctComb)` as `correctCom` is boolean

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+    desc.add<edm::InputTag>("PFJetSrc", edm::InputTag("hltAK4PFJetsCorrected"))->setComment("Input collection of PFJets");
+    desc.add<edm::InputTag>("TauSrc", edm::InputTag("hltSinglePFTau20TrackPt1LooseChargedIsolationReg"))
+      ->setComment("Input collection of PFTaus that have passed ID and isolation requirements");
+    desc.add<double>("extraTauPtCut", 45)->setComment("In case of asymmetric tau pt cuts");
```
Does it make sense to add a "regular" tauPt cut in case a threshold applied in the filtered tau collection is too loose? It can happen in case the same filtered taus are shared by different paths with different thresholds.

namppl replied:
DiXXX triggers usually have a subleading pt filter with minN = 2 and a leading pt filter and the module is supposed to take the filter product of the subleading pt filter. If paths use different subleading pt thresholds, they likely use different filters and "TauSrc" should be different for them. So I think it is okay to have only one tau pt cut in the module.

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+    edm::Handle<trigger::TriggerFilterObjectWithRefs> tauJets;
+    iEvent.getByToken(tauSrc_, tauJets);
+
+    edm::Handle<reco::PFJetCollection> PFJets;
```
Decapitalise: PFJets -> pfJets

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+    iEvent.getByToken(pfJetSrc_, PFJets);
+
+    trigger::VRpftau taus;
+    tauJets->getObjects(trigger::TriggerTau, taus);
```
I think you can simplify it as follows (no need use edm::Handle):
```
trigger::VRpftau taus;
iEvent.get(tauSrc_).objsWithRefs.getObjects(trigger::TriggerTau, taus);
```

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+    std::unique_ptr<reco::PFJetCollection> cleanedPFJets(new reco::PFJetCollection);
+
+    edm::Handle<trigger::TriggerFilterObjectWithRefs> tauJets;
+    iEvent.getByToken(tauSrc_, tauJets);
```
The `tauJets` handle is not needed, see comment to ll. 30-31. So ll. 24-25 can be removed.

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+      ->setComment("Input collection of PFTaus that have passed ID and isolation requirements");
+    desc.add<double>("extraTauPtCut", 45)->setComment("In case of asymmetric tau pt cuts");
+    desc.add<double>("mjjMin", 500)->setComment("VBF dijet mass condition");
+    desc.add<double>("Min_dR", 0.5)->setComment("Minimum dR outside of which PFJets will be saved");
```
Please use write more comprehensible comment, something as "Minimum dR between PFJet and filtered PFTau"

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+#include "RecoTauTag/HLTProducers/interface/PFJetsTauCorrelationCondition.h"
+#include "Math/GenVector/VectorUtil.h"
```
needed?

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+#include "RecoTauTag/HLTProducers/interface/PFJetsTauCorrelationCondition.h"
+#include "Math/GenVector/VectorUtil.h"
+#include "DataFormats/HLTReco/interface/TriggerTypeDefs.h"
+#include "FWCore/Utilities/interface/EDMException.h"
```
needed?

mbluj commented on RecoTauTag/HLTProducers/src/PFJetsTauCorrelationCondition.cc:
```diff
@@ -0,0 +1,86 @@
+//
+PFJetsTauCorrelationCondition::PFJetsTauCorrelationCondition(const edm::ParameterSet& iConfig)
+    : tauSrc_(consumes<trigger::TriggerFilterObjectWithRefs>(iConfig.getParameter<edm::InputTag>("TauSrc"))),
+      pfJetSrc_(consumes<reco::PFJetCollection>(iConfig.getParameter<edm::InputTag>("PFJetSrc"))),
```
It is not necessary to specify type in `consumes`, i.e. `consumes<...>(...)`->`consumes(...)`

mbluj commented on RecoTauTag/HLTProducers/interface/PFJetsTauCorrelationCondition.h:
```diff
@@ -0,0 +1,34 @@
+#include "DataFormats/JetReco/interface/CaloJetCollection.h"
+#include "DataFormats/TauReco/interface/PFTauFwd.h"
+#include "DataFormats/HLTReco/interface/TriggerFilterObjectWithRefs.h"
+#include "DataFormats/HLTReco/interface/TriggerObject.h"
```
Not needed

mbluj reviewed (changes requested):
Dear proponents,
thank you for this development. 

You can find my detailed comments inline. What concerns generalities:
* Could you try to change name of the module to more descriptive, please? This module is a producer of a collection of PFJets now overlapping with pairs of taus and forming pairs with a minimal invariant mass - could it be somehow codded in the name of the module?
* The module produces PFJets, but has "hidden" requirement that there are also at least two (filtered) PFTaus in the event - is it what you want? If so it is indeed not only producer, but also correlation condition checker. Anyway more descriptive name could be usefull.
*  Please add description what exactly the module does in its header file
*  Is there any reason to use  `global::EDProducer<>` instead of `stream::EDProducer<>`
* After implementing requested code changes please run `scram b code-fromat` (you can do it also before)
* Remove unnecessary include files, includes not needed in the header file move to the implementation file

namppl commented:
Dear Michal,

Many thanks for the detailed comments and sorry for the late response. I was not available last Friday. Your comments have been all implemented.

Best regards,
Kyungwook

namppl commented:
> * Could you try to change name of the module to more descriptive, please? This module is a producer of a collection of PFJets now overlapping with pairs of taus and forming pairs with a minimal invariant mass - could it be somehow codded in the name of the module?

I would change the name to "PFDiJetCorrCheckerWithDiTau"

> * The module produces PFJets, but has "hidden" requirement that there are also at least two (filtered) PFTaus in the event - is it what you want? If so it is indeed not only producer, but also correlation condition checker. Anyway more descriptive name could be usefull.

Yes, the module makes sense only if there are at least 2 jets and 2 taus. "DiJet" and "DiTau" in the new name hopefully clarify that point.

> * Please add description what exactly the module does in its header file

I'll add it asap.

> * Is there any reason to use  `global::EDProducer<>` instead of `stream::EDProducer<>`

No, the module processes event by event so it should be fine with stream. It's now stream::EDProducer.

> * After implementing requested code changes please run `scram b code-fromat` (you can do it also before)

Done

> * Remove unnecessary include files, includes not needed in the header file move to the implementation file

Done

mbluj commented on RecoTauTag/HLTProducers/src/PFDiJetCorrCheckerWithDiTau.cc:
```diff
@@ -0,0 +1,85 @@
+      ->setComment("Input collection of PFTaus that have passed ID and isolation requirements");
+  desc.add<double>("extraTauPtCut", 45)->setComment("In case of asymmetric tau pt cuts");
+  desc.add<double>("mjjMin", 500)->setComment("VBF dijet mass condition");
+  desc.add<double>("Min_dR", 0.5)->setComment("Minimum dR between PFJets and filtered PFTaus");
```
Min_dR -> dRmin

mbluj commented:
Thank you for implementing my changes. The missing thing is a short description (in header file) what the module does. I also spotted that name of configuration argument in l.80 has not been updated -please fix it too.
Finally, I would like to ask you to provide me a recipe (a cfg file) to run the module - I would like to check that it runs w/o failing, but of course I do not pretend to check if it works correctly.

namppl commented:
I've fixed L80 and added a description in the header file. I can get my colleague to provide a working version of HLT configuration but it will require some tedious works (merge PRs and tweaks on L1) to run the configuration. We've tested the code works correctly last week and will double-check the updated code also works fine.

mbluj commented:
> I've fixed L80 and added a description in the header file. I can get my colleague to provide a working version of HLT configuration but it will require some tedious works (merge PRs and tweaks on L1) to run the configuration. We've tested the code works correctly last week and will double-check the updated code also works fine.

Thank you.

Concerning the test configuration: I think you need dump HLT path from ConfDB to cfg, edit it manually to add the new module and provide me the cfg with instruction how to use it. It should be done to be sure that nothing is messed up, but it does not need any very realistic test scenario.

mbluj commented:
For records: test configuration was delivered and tests with it are successful => this PR is going to be merged.
